daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala

```diff
@@ -354,31 +364,43 @@ class Engine(val config: EngineConfig = Engine.StableConfig) {
             pkg => {
               compiledPackages.addPackage(pkgId, pkg).flatMap { _ =>
                 callback(compiledPackages)
-                interpretLoop(machine, time)
+                interpretLoop(machine, disclosures, time)
               }
             },
           )
 
-        case SResultNeedContract(contractId, _, callback) =>
-          return Result.needContract(
-            contractId,
-            { coinst =>
-              callback(coinst)
-              interpretLoop(machine, time)
-            },
-          )
+        case SResultNeedContract(contractId, _, callback) => {
+          def continueWithContract = (coinst: VersionedContractInstance) => {
+            callback(coinst)
+            interpretLoop(machine, disclosures, time)
+          }
+          disclosures.contractById.get(contractId) match {
+            case None =>
+              return Result.needContract(
+                contractId,
+                continueWithContract,
+              )
+            case Some(coinst) => return continueWithContract(coinst)
+          }
+        }
 
         case SResultNeedTime(callback) =>
           callback(time)
 
         case SResultNeedKey(gk, _, cb) =>
-          return ResultNeedKey(
-            gk,
-            { result =>
-              discard[Boolean](cb(result))
-              interpretLoop(machine, time)
-            },
-          )
+          def continueWithCoid = (result: Option[ContractId]) => {
+            discard[Boolean](cb(result))
+            interpretLoop(machine, disclosures, time)
+          }
+
+          disclosures.contractIdByKey.get(gk.globalKey.hash) match {
+            case None =>
+              return ResultNeedKey(
+                gk,
+                continueWithCoid,
+              )
+            case Some(coid) => return continueWithCoid(Some(coid))
```

Comment:
maybe add a TODO to validate that the key hash was correct here.

Reply:
added an issue and a comment.